reducer/src/main/java/com/graphicsfuzz/reducer/reductionopportunities/LiteralToUniformReductionOpportunity.java

```diff
@@ -16,27 +16,120 @@
+import com.graphicsfuzz.common.ast.IParentMap;
+import com.graphicsfuzz.common.ast.TranslationUnit;
+import com.graphicsfuzz.common.ast.decl.ArrayInfo;
+import com.graphicsfuzz.common.ast.decl.Declaration;
+import com.graphicsfuzz.common.ast.decl.VariableDeclInfo;
+import com.graphicsfuzz.common.ast.decl.VariablesDeclaration;
+import com.graphicsfuzz.common.ast.expr.ArrayIndexExpr;
+import com.graphicsfuzz.common.ast.expr.ConstantExpr;
+import com.graphicsfuzz.common.ast.expr.FloatConstantExpr;
+import com.graphicsfuzz.common.ast.expr.IntConstantExpr;
+import com.graphicsfuzz.common.ast.expr.VariableIdentifierExpr;
+import com.graphicsfuzz.common.ast.type.BasicType;
+import com.graphicsfuzz.common.ast.type.QualifiedType;
+import com.graphicsfuzz.common.ast.type.TypeQualifier;
 import com.graphicsfuzz.common.ast.visitors.VisitationDepth;
-import com.graphicsfuzz.common.util.PipelineInfo;
+import com.graphicsfuzz.common.transformreduce.ShaderJob;
+import com.graphicsfuzz.util.Constants;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
 
 public class LiteralToUniformReductionOpportunity
     extends AbstractReductionOpportunity {
 
-  private final Expr literalExpr;
-  private final PipelineInfo pipelineInfo;
+  private final ConstantExpr literalExpr;
+  private final TranslationUnit translationUnit;
+  private final ShaderJob shaderJob;
+  private final Number numericValue;
+  private final String arrayName;
+  private final BasicType basicType;
 
-  LiteralToUniformReductionOpportunity(Expr literalExpr,
-                                       PipelineInfo pipelineInfo,
+  LiteralToUniformReductionOpportunity(ConstantExpr literalExpr,
+                                       TranslationUnit translationUnit,
+                                       ShaderJob shaderJob,
                                        VisitationDepth depth) {
+
     super(depth);
+
     this.literalExpr = literalExpr;
-    this.pipelineInfo = pipelineInfo;
+    this.translationUnit = translationUnit;
+    this.shaderJob = shaderJob;
+
+    // If the uniform array doesn't exist in the pipeline info adds an empty array
+    if (literalExpr instanceof IntConstantExpr) {
+      this.arrayName = Constants.INT_LITERAL_UNIFORM_VALUES;
+      this.basicType = BasicType.INT;
+      this.numericValue = ((IntConstantExpr) literalExpr).getNumericValue();
+    } else {
+      this.arrayName = Constants.FLOAT_LITERAL_UNIFORM_VALUES;
+      this.basicType = BasicType.FLOAT;
+      this.numericValue = Float.valueOf(((FloatConstantExpr) literalExpr).getValue());
+    }
+
+    if (!shaderJob.getPipelineInfo().hasUniform(this.arrayName)) {
+      shaderJob.getPipelineInfo().addUniform(this.arrayName, this.basicType,
+          Optional.of(0), new ArrayList<>());
+      shaderJob.getPipelineInfo().addUniformBinding(this.arrayName, false,
+          shaderJob.getPipelineInfo().getNumUniforms());
+    }
```

Comment:
This does not belong in the constructor.  We do not want to mutate the shader job when we create this opportunity - any mutations to the shader job should occur only if and when the opportunity is actually applied.

Reply:
Moved to applyReductionImpl.